### DISCUSSION TITLE
How to update docker tags like `myimage:sha-<short-git-sha>`

### DISCUSSION THREAD

**Discussion by felipecrs** [renovatebot/renovate#27339] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.192.1

### If you're self-hosting Renovate, select which platform you are using.

None

### What is your question?

I have a docker image that isn't versioned with semver. Instead, I push tags like `:sha-<short-git-sha>`.

I am trying to make Renovate propose upgrades for it, but so far no success.

I've tried `versioning: loose` and `versioning: git`, both with `extractVersion=^sha-(?<version>.+)$`.

When using `git`:

```json
               {
                 "depName": "myregistry.com/myrepo/cicd",
                 "currentValue": "053ea93",
                 "datasource": "docker",
                 "versioning": "git",
                 "extractVersion": "^sha-(?<version>.+)$",
                 "replaceString": "# renovate: datasource=docker depName=myregistry/myrepo/cicd versioning=git extractVersion=^sha-(?<version>.+)$\nARG CICD_VERSION=\"053ea93\"\n",
                 "updates": [],
                 "packageName": "myregistry/myrepo/cicd",
                 "warnings": [],
                 "registryUrl": "https://myregistry.com",
                 "currentVersion": "053ea93",
                 "fixedVersion": "053ea93"
               },
```

And when using `loose` I receive a `skipReason: invalid-value`.

Both cases, Renovate can't find any update. I see in the logs:

```console
DEBUG: getLabels(https://myregistry.com, myrepo/cicd, latest) (repository=local)
DEBUG: getManifestResponse(https://myregistry.com, myrepo/cicd, latest, get) (repository=local)
DEBUG: found labels in manifest (repository=local)
       "labels": {
         "org.opencontainers.image.version": "sha-0bf3039"
       }
DEBUG: Found no satisfying versions with 'git' versioning (repository=local)
```

So, I was expecting that Renovate would extract the version out of the label for the `:latest` tag and propose that as an update.

Am I missing something?

**Comment by viceice** [comment, marked as answer]
not possible because those sha strings don't have an order and the docker API doesn't provide release dates

**Reply by felipecrs** [reply]
I see. That makes sense.

Would you be welcome for a fallback mode where Renovate: 

1. Reads the labels for the `:latest` tag
2. Uses the content of the label `org.opencontainers.image.version` as the latest version
3. If it differs from `currentVersion`, propose it as update.

?

**Reply by viceice** [reply]
that's not easily possible with current renovate architecture

**Reply by felipecrs** [reply]
I see. Thank you.

**Reply by rarkins** [reply]
Maybe using followTag? Buy this sounds like a proprietary solution  so we prefer to avoid those

**Reply by felipecrs** [reply]
That's interesting. I will try it out.

But why proprietary? I think it makes a lot of sense. The OCI labels are kind of a standard by now.

**Reply by rarkins** [reply]
Using a standard label in a proprietary manner is still a proprietary solution

**Reply by felipecrs** [reply]
Sorry, I don't get why the word proprietary fits here. But I don't think it's important anyway, lol.

But, @rarkins, did you mean that we could implement followTag for docker datasource for my use case or did you mean that it could work today?

Here is what I tried:

```json
{
  "packageRules": [
    {
      "matchPackagePatterns": ["*"],
      "enabled": false
    },
    {
      "matchPackageNames": ["myregistry.com/myrepo/cicd"],
      "enabled": true,
      "followTag": "latest"
    }
  ]
}
```

Then, I see in logs:

```
DEBUG: getLabels(https://myregistry.com, myrepo/cicd, latest) (repository=local)
DEBUG: getManifestResponse(https://myregistry.com, myrepo/cicd, latest, get) (repository=local)
DEBUG: found labels in manifest (repository=local)
       "labels": {
         "org.opencontainers.image.version": "sha-0bf3039"
       }
```

Later:

```json
               {
                 "depName": "myregistry.com/myrepo/cicd",
                 "currentValue": "b60f3d4",
                 "datasource": "docker",
                 "versioning": "git",
                 "extractVersion": "^sha-(?<version>.+)$",
                 "replaceString": "# renovate: datasource=docker depName=myregistry.com/myrepo/cicd versioning=git extractVersion=^sha-(?<version>.+)$\nARG CICD_VERSION=\"b60f3d4\"\n",
                 "updates": [],
                 "packageName": "myregistry.com/myrepo/cicd",
                 "warnings": [
                   {
                     "topic": "myregistry.com/myrepo/cicd",
                     "message": "Can't find version with tag latest for docker package myregistry.com/myrepo/cicd"
                   }
                 ],
                 "registryUrl": "https://myregistry.com"
               },
```

I see no other indicator in the logs of what's going on.

**Reply by rarkins** [reply]
I mean it could be reused and not that it already works. However we wouldn't want to support it unless it was a standard and documented docker metadata approach

**Reply by felipecrs** [reply]
Perfect. Understood now what you mean by proprietary.

Honestly, I think it's a very reasonable approach:

1. You have a tag named `latest`
2. You need to find out what is the version it points to
   - You can use the standard OCI label `org.opencontainers.image.version` if present in such `latest` tag to find out the version

I am not aware of any documentation around it. And I would not be surprised if there was never a need for it in the docker world.

I will try to come up with a PR, then we can continue from there.
